File: compiler/src/utils/span.rs
use core::ops::Range;

#[derive(Default, Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct Span {
    pub start: usize,
    pub end: usize,
}

impl Span {
    pub fn new(start: usize, end: usize) -> Self {
        assert!(start <= end, "Span start must be <= end");
        Self { start, end }
    }

    pub const fn len(&self) -> usize {
        self.end - self.start
    }

    pub const fn is_empty(&self) -> bool {
        self.start >= self.end
    }

    pub const fn contains(&self, pos: usize) -> bool {
        pos >= self.start && pos < self.end
    }

    pub const fn contains_span(&self, other: Span) -> bool {
        other.start >= self.start && other.end <= self.end
    }

    pub fn to(self, other: Span) -> Self {
        Self::new(
            std::cmp::min(self.start, other.start),
            std::cmp::max(self.end, other.end),
        )
    }
}

/// let span: Span = (0..10).into();
impl From<Range<usize>> for Span {
    fn from(range: Range<usize>) -> Self {
        Self::new(range.start, range.end)
    }
}

///  &str[span.into()]
impl From<Span> for Range<usize> {
    fn from(span: Span) -> Self {
        span.start..span.end
    }
}

File: compiler/utils/span.zig
const std = @import("std");
const assert = std.debug.assert;

pub const Span = struct {
    start: usize,
    end: usize,

    pub fn new(start: usize, end: usize) Span {
        assert(start <= end);
        return .{
            .start = start,
            .end = end,
        };
    }

    pub fn len(self: Span) usize {
        return self.end - self.start;
    }

    pub fn merge(self: Span, other: Span) Span {
        return .{
            .start = @min(self.start, other.start),
            .end = @max(self.end, other.end),
        };
    }

    /// 从源码中切出对应的字符串
    pub fn slice(self: Span, source: []const u8) []const u8 {
        assert(self.end <= source.len);
        return source[self.start..self.end];
    }
};
